File: packages/ocr-onnx/benchmarks/quality_eval/metrics/wer.py
"""Word Error Rate (WER) metric for OCR evaluation."""

from typing import List


def levenshtein_distance_words(words1: List[str], words2: List[str]) -> int:
    """Compute Levenshtein distance at word level.

    Args:
        words1: First list of words
        words2: Second list of words

    Returns:
        Integer word-level edit distance
    """
    if len(words1) > len(words2):
        words1, words2 = words2, words1

    distances = list(range(len(words1) + 1))
    for i2, w2 in enumerate(words2):
        distances_ = [i2 + 1]
        for i1, w1 in enumerate(words1):
            if w1 == w2:
                distances_.append(distances[i1])
            else:
                distances_.append(1 + min(distances[i1], distances[i1 + 1], distances_[-1]))
        distances = distances_
    return distances[-1]


def compute_wer(prediction: str, reference: str, normalize: bool = True) -> float:
    """Compute Word Error Rate (WER).

    WER = word_edit_distance(prediction, reference) / word_count(reference)

    Args:
        prediction: OCR output text
        reference: Ground truth text
        normalize: If True, normalize texts before comparison

    Returns:
        WER value (0.0 = perfect match, higher = more errors)
    """
    if not isinstance(prediction, str):
        prediction = str(prediction) if prediction is not None else ""
    if not isinstance(reference, str):
        reference = str(reference) if reference is not None else ""

    if normalize:
        prediction = prediction.lower().strip().replace("\n", " ")
        reference = reference.lower().strip().replace("\n", " ")

    pred_words = prediction.split()
    ref_words = reference.split()

    if len(ref_words) == 0:
        return 1.0 if len(pred_words) > 0 else 0.0

    distance = levenshtein_distance_words(pred_words, ref_words)
    return distance / len(ref_words)

File: packages/bci-whispercpp/lib/wer.js
'use strict'

/**
 * Compute Word Error Rate between hypothesis and reference.
 * Uses Levenshtein distance on word sequences.
 * @param {string} hypothesis
 * @param {string} reference
 * @returns {number} WER as a ratio (0.0 = perfect, 1.0 = 100% errors)
 */
function computeWER (hypothesis, reference) {
  const hyp = hypothesis.toLowerCase().trim().split(/\s+/).filter(Boolean)
  const ref = reference.toLowerCase().trim().split(/\s+/).filter(Boolean)

  if (ref.length === 0) return hyp.length === 0 ? 0 : 1

  const n = ref.length
  const m = hyp.length
  const dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0))

  for (let i = 0; i <= n; i++) dp[i][0] = i
  for (let j = 0; j <= m; j++) dp[0][j] = j

  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      if (ref[i - 1] === hyp[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1]
      } else {
        dp[i][j] = 1 + Math.min(
          dp[i - 1][j],
          dp[i][j - 1],
          dp[i - 1][j - 1]
        )
      }
    }
  }

  return dp[n][m] / n
}

module.exports = { computeWER }
